astroid/decorators.py

```diff
@@ -84,28 +84,22 @@ def path_wrapper(func):
     def wrapped(node, context=None, _func=func, **kwargs):
         """wrapper function handling context"""
         if context is None:
-            context = contextmod.InferenceContext()
-        if context.push(node):
-            return None
+            context = contextmod.global_context
 
-        yielded = set()
-        generator = _func(node, context, **kwargs)
-        try:
-            while True:
-                res = next(generator)
-                # unproxy only true instance, not const, tuple, dict...
-                if res.__class__.__name__ == "Instance":
-                    ares = res._proxied
-                else:
-                    ares = res
-                if ares not in yielded:
-                    yield res
-                    yielded.add(ares)
-        except StopIteration as error:
-            if error.args:
-                return error.args[0]
+        execution = context.push(node)
+        if execution.is_recursing():
             return None
 
+        execution.increment()
+        inferred = execution.inferred
+        if inferred:
+            yield from list(inferred)
+        else:
+            generator = _func(node, context, **kwargs)
+            for res in generator:
+                yield res
+                inferred.add(res)
```

Comment:
I might be wrong here ... but say "generator" has length 5. If `next` is only called once, then `execution.inferred` will only have 1 element, not 5 - which will then be reused next time as a single item iterator, and the remaining 4 items will be forever lost. Then again ... I've only got 3 failed tests, and fixing this doesn't seem to help the one I'm focusing on.